include/arch/arm/aarch64/lib_helpers.h

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2021 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_ARCH_ARM_AARCH64_LIB_HELPERS_H_
+#define ZEPHYR_INCLUDE_ARCH_ARM_AARCH64_LIB_HELPERS_H_
+
+#ifndef _ASMLANGUAGE
+
+#include <arch/arm/aarch64/cpu.h>
+#include <stdint.h>
+
+#define read_sysreg(reg)						\
+({									\
+	uint64_t val;							\
+	__asm__ volatile ("mrs %0, " STRINGIFY(reg) "\n\t"		\
+			  : "=r" (val) :: "memory");			\
+	val;								\
+})
+
+#define write_sysreg(val, reg)						\
+({									\
+	__asm__ volatile ("msr " STRINGIFY(reg) ", %0\n\t"		\
+			  :: "r" (val) : "memory");			\
+})
+
+
+#define MAKE_REG_HELPER(reg)						\
+	static ALWAYS_INLINE uint64_t read_##reg(void)			\
+	{								\
+		return read_sysreg(reg);				\
+	}								\
+	static ALWAYS_INLINE void write_##reg(uint64_t val)		\
+	{								\
+		write_sysreg(val, reg);					\
+	}
+
+#define MAKE_REG_HELPER_EL123(reg) \
+	MAKE_REG_HELPER(reg##_el1) \
+	MAKE_REG_HELPER(reg##_el2) \
+	MAKE_REG_HELPER(reg##_el3)
+
+MAKE_REG_HELPER(cntv_ctl_el0)
+MAKE_REG_HELPER(cntv_cval_el0)
+MAKE_REG_HELPER(cntvct_el0);
+MAKE_REG_HELPER(currentel);
+MAKE_REG_HELPER(daif)
+
+MAKE_REG_HELPER_EL123(actlr)
+MAKE_REG_HELPER_EL123(cpacr)
+MAKE_REG_HELPER_EL123(cptr)
+MAKE_REG_HELPER_EL123(elr)
+MAKE_REG_HELPER_EL123(esr)
+MAKE_REG_HELPER_EL123(far)
+MAKE_REG_HELPER_EL123(mair)
+MAKE_REG_HELPER_EL123(scr)
+MAKE_REG_HELPER_EL123(sctlr)
+MAKE_REG_HELPER_EL123(spsr)
+MAKE_REG_HELPER_EL123(tcr)
+MAKE_REG_HELPER_EL123(ttbr0)
+MAKE_REG_HELPER_EL123(vbar)
+
+static ALWAYS_INLINE void enable_debug_exceptions(void)
+{
+	__asm__ volatile ("msr DAIFClr, %0\n\t"
+			  :: "i" (DAIFSET_DBG_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void disable_debug_exceptions(void)
+{
+	__asm__ volatile ("msr DAIFSet, %0\n\t"
+			  :: "i" (DAIFSET_DBG_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void enable_serror_exceptions(void)
+{
+	__asm__ volatile ("msr DAIFClr, %0\n\t"
+			  :: "i" (DAIFSET_ABT_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void disable_serror_exceptions(void)
+{
+	__asm__ volatile ("msr DAIFSet, %0\n\t"
+			  :: "i" (DAIFSET_ABT_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void enable_irq(void)
+{
+	__asm__ volatile ("msr DAIFClr, %0\n\t"
+			  :: "i" (DAIFSET_IRQ_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void disable_irq(void)
+{
+	__asm__ volatile ("msr DAIFSet, %0\n\t"
+			  :: "i" (DAIFSET_IRQ_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void enable_fiq(void)
+{
+	__asm__ volatile ("msr DAIFClr, %0\n\t"
+			  :: "i" (DAIFSET_FIQ_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void disable_fiq(void)
+{
+	__asm__ volatile ("msr DAIFSet, %0\n\t"
+			  :: "i" (DAIFSET_FIQ_BIT) : "memory");
+}
+
+#define dsb()	__asm__ volatile ("dsb sy" ::: "memory")
+#define dmb()	__asm__ volatile ("dmb sy" ::: "memory")
+#define isb()	__asm__ volatile ("isb" ::: "memory")
```

Comment:
This needs a compiler barrier. (As does `dsb` likely, as `dsb` is a superset of `isb`).

`isb` is typically used when previous _instructions_ need to complete. The `memory` clobber here only tells the compiler that it cannot reorder _memory_ accesses around this instruction, not generic operations.

For a simple example where that shows up (although the rearrangement here is very likely benign):

```c
float foo(float a, float b) {
    a /= b;
    isb();
    return a;
}
```
...which turns into

```asm
foo(float, float):
        isb
        fdiv    s0, s0, s1
        ret
```
https://godbolt.org/z/b17dnd

(Imagine for a moment that that this was the result of inlining or something and that isb was there to ensure that the division-by-zero trap was disabled...)

...unfortunately, now that I look GCC doesn't have a compiler scheduler barrier, which is unfortunate.

So I strongly suggest adding a comment in this file that all of these macros need a `memory` clobber, at the very least, lest they get rearranged around isb.

Reply:
Yup. Comment added.

Reply:
...Hmm, I don't see a comment here?

Reply:
hahah, I added it on top https://github.com/zephyrproject-rtos/zephyr/blob/1609e58122518395d1e59381f6f419e933dcf820/include/arch/arm/aarch64/lib_helpers.h#L15